gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,162 @@
+package com.google.cloud.hadoop.fs.gcs;
+
+import static com.google.api.client.util.Preconditions.checkNotNull;
+
+import com.google.api.client.auth.oauth2.ClientParametersAuthentication;
+import com.google.api.client.auth.oauth2.RefreshTokenRequest;
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageOptions;
+import com.google.cloud.hadoop.util.AccessTokenProvider;
+import com.google.cloud.hadoop.util.HttpTransportFactory;
+import com.google.cloud.hadoop.util.RedactedString;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+public class RefreshTokenAuth2Provider implements AccessTokenProvider {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private static final AccessToken EXPIRED_TOKEN = new AccessToken("", -1L);
+  private static DateTimeFormatter dateFormat =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS").withZone(ZoneId.systemDefault());
+  private Configuration config;
+  private AccessToken accessToken = EXPIRED_TOKEN;
+  private HttpTransport httpTransport;
+  private Optional<RedactedString> previousRefreshToken = Optional.empty();
+
+  @Override
+  public AccessToken getAccessToken() {
+    return this.accessToken;
+  }
+
+  @Override
+  public void refresh() {
+    if (logger.atFine().isEnabled()) {
+      String expireAt =
+          accessToken.getExpirationTimeMilliSeconds() == null
+              ? "NEVER_EXPIRE"
```

Comment:
Is this case actually possible? I don't think there's any valid scenario where the expiration time should be null.

I'd rather add a precondition check to the AccessToken class that the expiration time is not null.

Reply:
Alternatively, if it is null, then treat it like the token already expired.

Reply:
According to the spec, the `expires_in` can be null:

https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.2
```
         RECOMMENDED.  The lifetime in seconds of the access token.  For
         example, the value "3600" denotes that the access token will
         expire in one hour from the time the response was generated.
         If omitted, the authorization server SHOULD provide the
         expiration time via other means or document the default value.
```

In the case of Google, it seems always set so I guess we can assert it is present

Reply:
What I will do, is as you said, set it up as expired and print a warning.
Even if today the Google OAuth2 provider doesn't document that this value can be null, it may in the future and this warning will help debugging the situation.